File: euler/p17.c
#include <stdio.h>
#include <stdint.h>

uint64_t chars(int n) {
  // zero one two three four five six seven eight nine
  int unites[10] = {0, 3, 3, 5, 4, 4, 3, 5, 5, 4};

  // zero ten twenty thirty fourty fifty sixty seventy eighty ninety
  int dizaines[10] = {0, 3, 6, 6, 6, 5, 5, 7, 6, 6};

  // ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen
  int sadique[10] = {3, 6, 6, 8, 8, 7, 7, 9, 8, 8};

  uint64_t count = 0;
  
  int unite = n % 10;
  int dizaine = ((n - unite) / 10) % 10;
  int centaine = ((n - dizaine * 10 - unite) / 100) % 100;

  if (centaine != 0) {
    // "hundred"
    count += unites[centaine] + 7;

    // "and"
    if (dizaine != 0 || unite != 0) {
      count += 3;
    }
  }

  if (dizaine == 1) {
    count += sadique[dizaine];
  } else {
    count += dizaines[dizaine] + unites[unite];
  }
  
  return count;
}

int main() {
  uint64_t count = 0;

  for (int i = 1; i < 1000; i++) {
    count += chars(i);
  }

  // one thousand
  count += 11;

  printf("Compte: %lu\n", count);
}

File: euler/p17.ml

let somme n =
  let unites = [| 0; 3; 3; 5; 4; 4; 3; 5; 5; 4|] in
  let dizaines = [| 0; 3; 6; 6; 6; 5; 5; 7; 6; 6 |] in
  let milliers = [| 0; 11 |] in
  let sadique = [| 3; 6; 6; 8; 8; 7; 7; 9; 8; 8 |] in
  let chars n =
    let unite = n mod 10 in
    let dizaine = ((n - unite) / 10) mod 10 in
    let centaine = ((n - (n mod 100)) / 100) mod 10 in
    let millier = ((n - (n mod 1000)) / 1000) mod 10 in
    (* Calcul *)
    milliers.(millier)(*  Millier *)
    + (if centaine <> 0 then unites.(centaine) + 7  else 0)(* Centaine *)
    + (if centaine <> 0 && (dizaine <> 0 || unite <> 0) then 3 else 0)(* And *)
    + (if dizaine = 1 then sadique.(dizaine) else 0)(* Dizaine *)
    + (if dizaine <> 1 then dizaines.(dizaine) + unites.(unite) else 0)(* Unité *)
  in
  let s = ref 0 in
  for i = 1 to n do
    s := !s + chars i;
  done;
  !s
